### PR TITLE
fake commit

### PR DESCRIPTION
**Pull request by Bnjwmn**

### CODE REVIEW

File: src2/main/include/Climb.h [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+enum ControllerInput{
+  A,
+  B,
+  X,
+  Y,
+  RB,
+  LB,
+};
+
```
Comment (sunaynas) [inline src2/main/include/Climb.h:36]: why is controller input in climb?

File: src2/main/include/Robot.h [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+  frc::SendableChooser<std::string> m_chooser;
+  const std::string kAutoNameDefault = "Default";
+  const std::string kAutoNameCustom = "My Auto";
+  std::string m_autoSelected;
+
+  Climb m_climb{"m_climb", true, false};
+
+  ControllerInput Input;
+
+  frc::XboxController m_controller{ClimbIDs::CLIMB_CONTROLLER};
```
Comment (sunaynas) [inline src2/main/include/Robot.h:43]: You should use [caleb's controller class](https://docs.google.com/document/d/1GWZ8jY3yyM1fjt0oaqTTqf8iZJqo__Anjgq_FBxjAf4/edit) its much cleaner and imo rlly nice (even if the docs are not)

File: src2/main/include/ClimbConstants.h [hunk 3]
```diff
@@ -0,0 +1,14 @@
+#pragma once
+
+namespace ClimbIDs {
+    const int CLIMB_MOTOR=31;
+    const int CLIMB_CONTROLLER=2;
```
Comment (sunaynas) [inline src2/main/include/ClimbConstants.h:5]: this should be in smt like controller constants ([ex](https://github.com/Eaglestrike/2024-RobotCode/blob/main/src/main/include/Constants/ControllerConstants.h)) since its not like you get a whole separate controller for climb. If you use caleb's controller class and our old controller constants itll take care of all of this stuff for u

File: src2/main/include/ClimbConstants.h [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+    const int BRAKE=41;
+}
+namespace ClimbConstants{
+const double Total_Length= 22.0;//in inches
+const double Spindle_Radius= 0.5;//in inches 
+const double Pi=3.14;
+const double Max_Voltage= 2;//in volts
+const double Accel_Voltage_Max=0.002;//in volts/s
```
Comment (sunaynas) [inline src2/main/include/ClimbConstants.h:13]: consts should be in screaming snake case (i.e. TOTAL_LENGTH)

File: src2/main/cpp/Robot.cpp [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+  if(m_controller.GetYButtonPressed()){
+    return Y;
+  }
+  if(m_controller.GetLeftBumperPressed()){
+    return LB;
+  }
+  if(m_controller.GetRightBumperPressed()){
+    return RB;
+  }
+}
```
Comment (sunaynas) [inline src2/main/cpp/Robot.cpp:29]: Again, controller should be separate from stuff like robot and climb and should be wrapped in its own class

File: src2/main/cpp/Robot.cpp [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+    // Custom Auto goes here
+  } else {
+    // Default Auto goes here
+  }
+}
+
+void Robot::TeleopInit() {}
+
+void Robot::TeleopPeriodic() {
+m_climb.ClimbTeleopPeriodic(GetButtonInput(),m_controller.GetLeftY());
```
Comment (sunaynas) [inline src2/main/cpp/Robot.cpp:82]: Usually periodic calls don't take any parameters and are just continuously called. The internal state machine will handle if periodic will do anything at all.

Then here, in robot you'll have an instance of controller and climb and you'll use the instance of controller to check if buttons are pressed and then conditionally call action functions like climb.stow or something like that, which change the internal state of climb. Changing the internal state changes what the periodic call of climb will do when it is called.

So basically actions should be separated from the periodic function and controller logic should be handled here and not within climb because controller has nothing to do with climb and therefore its bad practice to put controller stuff in climb.

File: src2/main/include/Climb.h [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+#include <ctre/phoenix/motorcontrol/can/WPI_TalonSRX.h>
+
+using ctre::phoenix6::hardware::TalonFX;
+using ctre::phoenix::motorcontrol::can::WPI_TalonSRX;
+
+enum ClimbLocation{
+    AT_BOTTOM,
+    AT_TOP,
+    BETWEEN,
+};
```
Comment (sunaynas) [inline src2/main/include/Climb.h:15]: why do you need a state for the location? you already have the position

File: src2/main/include/Climb.h [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+
+enum ClimbState{
+    MANUAL,
+    AUTO,
```
Comment (sunaynas) [inline src2/main/include/Climb.h:19]: ``
```suggestion
    MOVING,
    AT_TARGET
```

File: src2/main/include/Climb.h [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+};
+
+enum ClimbTarget{
+    DOWN,
+    UP,
+    NONE,
```
Comment (sunaynas) [inline src2/main/include/Climb.h:25]:
```suggestion
```

File: src2/main/cpp/Climb.cpp [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+            break;
+        case LB:
+            RemoveBrake();
+            break;
+        case RB:
+            Brake();
+            break;
+        default:
+            break;
+    }
```
Comment (sunaynas) [inline src2/main/cpp/Climb.cpp:65]: this should all be handled in robot. toggle state, zero, removebrake, and brake should all be public functions that robot can call, and you should create 2 new methods called like stow and extend or something for case x&y which seem to handle changing the target. this is also where unbreak should be called and the state should be set to moving.

File: src2/main/cpp/Climb.cpp [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+        case DOWN:
+            MoveDown();
+            break;
+        default:
+            break;
+    }
+
+    if(m_state==MANUAL){
+        Manual(StickInput);
+    }
```
Comment (sunaynas) [inline src2/main/cpp/Climb.cpp:81]: you should check this first and then return after you do the manual stuff so the regular stuff doesnt run when ur in manual mode

File: src2/main/cpp/Climb.cpp [hunk 12]
```diff
@@ -0,0 +1,175 @@
...
+    
+    switch (m_target)
+    {
+        case UP:
+            MoveUp();
+            break;
+        case DOWN:
+            MoveDown();
+            break;
```
Comment (sunaynas) [inline src2/main/cpp/Climb.cpp:74]: this should be a switch on the state of the robot. If you are moving towards the target but not quite at the target, then if your target is up move up. if its down, move down. if you are at the target, do nothing and chill. while youre moving toward the target you should be checking if youre at the target and changing the state accordingly.

File: src2/main/cpp/Climb.cpp [hunk 13]
```diff
@@ -0,0 +1,175 @@
...
+        default:
+            break;
```
Comment (sunaynas) [inline src2/main/cpp/Climb.cpp:76]: also if default is doing nothing you dont need it

File: src2/main/cpp/Climb.cpp [hunk 14]
```diff
@@ -0,0 +1,175 @@
...
+    }
+    double Manual_Voltage = ClimbConstants::Max_Voltage*ManualInput;
+    Manual_Voltage = std::clamp(Manual_Voltage, -ClimbConstants::Max_Voltage, ClimbConstants::Max_Voltage);
+    m_motor.SetVoltage(units::volt_t(Manual_Voltage));
+}
+
+void Climb::MoveUp(){
+     if (m_location==AT_TOP){
+        return;
+     }
```
Comment (sunaynas) [inline src2/main/cpp/Climb.cpp:99]:
```suggestion
     if (m_pos>=ClimbConstants::Total_Length){
        m_state = AT_TARGET;
        return;
     }
```

File: src2/main/cpp/Climb.cpp [hunk 15]
```diff
@@ -0,0 +1,175 @@
...
+     if(m_braking){
+        RemoveBrake();
+        m_timer=0;
+        m_braking=false;
+     }
+     if(m_timer<0.24){
+        m_timer+=0.2;
+        return;
+     }
```
Comment (sunaynas) [inline src2/main/cpp/Climb.cpp:108]: should be done if state == AT_TARGET in state machine in teleopperiodic. also this brake logic wont work Im p sure. when you change the target (for example in the method stow or extend) you should call unbreak then start the timer and then in the switch statement while ur in the state MOVING you should keep updating timer and checking if it runs out and if its out only then you move. this is so that the gear isnt pressing against the rachet while its trying to disengage.

File: src2/main/cpp/Climb.cpp [hunk 16]
```diff
@@ -0,0 +1,175 @@
...
+        else if(m_voltage>0){
+            m_voltage-=ClimbConstants::Accel_Voltage_Max;
+            m_motor.SetVoltage(units::volt_t(m_voltage));
+            m_position=((m_motor.GetPosition().GetValueAsDouble())*(ClimbConstants::Spindle_Radius));
+            if(m_position==ClimbConstants::Total_Length){
+                m_location=AT_TOP;
+                Brake();
+                m_braking=true;
+            }
+        }
```
Comment (sunaynas) [inline src2/main/cpp/Climb.cpp:128]: position updating should be done in periodic. also you really dont need to increase/decrease the voltage like that sicne that would decrease precision. in general you shouldnt need m_voltage except for setting it for manual mode. you can just have motor.setvoltage(+or- ClimbConstants::MAX_VOLTAGE)

File: src2/main/cpp/Climb.cpp (hunk 11 above)
Comment (sunaynas) [reply src2/main/cpp/Climb.cpp:81]: maybe you can seperate periodic calls into manual periodic(double stick) and teleop periodic() and then use get state in robot to differentiate when to call

File: src2/main/cpp/Climb.cpp [hunk 17]
```diff
@@ -0,0 +1,175 @@
...
+        else if(m_voltage>0){
+            m_voltage-=ClimbConstants::Accel_Voltage_Max;
+            m_motor.SetVoltage(units::volt_t(-m_voltage));
+            m_position=((m_motor.GetPosition().GetValueAsDouble())*(ClimbConstants::Spindle_Radius));
+            if(m_position==ClimbConstants::Total_Length){
+                m_location=AT_BOTTOM;
+                Brake();
+                m_braking=true;
+            }
+        }
```
Comment (sunaynas) [inline src2/main/cpp/Climb.cpp:164]: same feedback as moveup

File: src2/main/include/Climb.h [hunk 18]
```diff
@@ -0,0 +1,62 @@
...
+        void MoveUp();
+        void MoveDown();
+        void Brake();
+        void RemoveBrake();
+        void ClimbTeleopPeriodic(ControllerInput Input, double StickInput);
+    private:
+        TalonFX m_motor{ClimbIDs::CLIMB_MOTOR}; 
+        double m_voltage=0; 
+        double m_position;
+        double m_targ_position;
```
Comment (sunaynas) [inline src2/main/include/Climb.h:55]: idt u need targ position if u alr have target

**Review by sunaynas — COMMENTED** [review]
overview:
1. fix state machine
2. fix brake logic
3. fix moving to target logic
4. use controller class and remove controller stuff from climb and places it doesnt need to be
5. clean up code
